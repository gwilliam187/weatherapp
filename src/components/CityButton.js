import React from 'react';
import { connect } from 'react-redux';

import { selectCity } from '../actions';
import { updateCitiesErrorMessage } from '../actions'

export class CityButton extends React.Component {
	
	handleOnClick = () => {
		this.props.updateCitiesErrorMessage(null, null);
		this.props.selectCity({
			cityName: this.props.cityName, 
			countryCode: this.props.countryCode
		});
	}

	render() {
		return (
			<div className='py-2'>
				<button 
					className={'btn btn-outline-secondary btn-lg city-button'} 
					onClick={this.handleOnClick}>
					{this.props.cityName}
				</button>
			</div>
		);
	}
}

const mapDispatchToProps = {
	selectCity, 
	updateCitiesErrorMessage
}

export default connect(
	null,
	mapDispatchToProps
)(CityButton);