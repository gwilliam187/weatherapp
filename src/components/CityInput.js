import React from 'react';
import { connect } from 'react-redux';

import { selectCity } from '../actions';
import { updateCitiesErrorMessage } from '../actions';

export class CityInput extends React.Component {
	state = {
		term: ''
	};

	handleOnKeyDown = (e) => {
		if(e.key === 'Enter') {
			this.props.updateCitiesErrorMessage(null, null);
			const termList = this.state.term.split(',');
			if(termList.length === 2) {
				this.props.selectCity({
					cityName: termList[0], 
					countryCode: termList[1]
				});
			}	else {
				this.props.updateCitiesErrorMessage(null, 'INVALID_INPUT');
			}

			this.setState({term: ''});
		}
	};

	handleOnChange = (e) => {
		this.setState({term: e.target.value});
	};

	render() {
		return(
			<div className='card col-12'>
				<div className='card-body'>
					<label>Add city</label>
					<input 
						type='text'
						placeholder='e.g. Berlin,de'
						value={this.state.term}
						onChange={this.handleOnChange}
						onKeyDown={this.handleOnKeyDown} 
						className='form-control form-control-lg'/>
					<div className='text-left text-danger mt-1'>
						{this.props.citiesErrorMessage}
					</div>
				</div>
			</div>
		);
	}
}

const mapStateToProps = state => {
	return {
		citiesErrorMessage: state.citiesErrorMessage
	};
};

const mapDispatchToProps = {
	selectCity, 
	updateCitiesErrorMessage
}

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(CityInput);