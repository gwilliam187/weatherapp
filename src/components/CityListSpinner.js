import React, { Component } from 'react';
import { connect } from 'react-redux';

class CityListSpinner extends Component {
	render() {
		if(this.props.citiesFetcher.isFetching) {
			return (
				<div className='text-primary'>
					<div className="spinner-border spinner-border-sm text-primary mr-2" role="status"></div>
					Fetching
				</div>
			);
		} else {
			return <div></div>
		}
	}
};

const mapStateToProps = state => ({
	citiesFetcher: state.citiesFetcher,
});

const mapDispatchToProps = {

};

export default connect(mapStateToProps, mapDispatchToProps)(CityListSpinner);