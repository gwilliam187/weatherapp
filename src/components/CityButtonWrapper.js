import React from 'react';
import CityButton from './CityButton';
import './custom-style.css';

class CityButtonWrapper extends React.Component {
	handleEvent = (cityName) => {
		this.props.eventHandler(cityName);
	}

	render() {
		return (
			<div className='container city-button-wrapper'>
				<div className='row header'>
					<div className='col-12'>
						<h1>Please choose a city</h1>
					</div>
				</div>
				<div className='row content'>
					<CityButton cityName='Berlin' countryCode='de' eventHandler={this.handleEvent} />
					<CityButton cityName='Potsdam' countryCode='de' eventHandler={this.handleEvent} />
					<CityButton cityName='Soest' countryCode='de' eventHandler={this.handleEvent} />
					<CityButton cityName='Jakarta' countryCode='id' eventHandler={this.handleEvent} />
					<CityButton cityName='Bandung' countryCode='id' eventHandler={this.handleEvent} />
					<CityButton cityName='Moscow' countryCode='ru' eventHandler={this.handleEvent} />
					<CityButton cityName='Bangkok' countryCode='th' eventHandler={this.handleEvent} />
				</div>
			</div>
		);
	}
}

export default CityButtonWrapper;