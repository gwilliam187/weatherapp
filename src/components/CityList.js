import React from 'react';
import { connect } from 'react-redux';

import CityInput from './CityInput';
import CityButton from './CityButton';

class CityList extends React.Component {
	renderList() {
		if(this.props.cities.length > 0) {
			return this.props.cities.map(city => {
				return (
					<CityButton city={ city } key={ city._id } />
				);
			});
		} else {
			return (
				<div className='default'>City list empty</div>
			);
		}
	}

	render() {
		return (
			<div className='col-12 card city-list'>
				<div className='card-body'>
					<div style={ styles.headerWrapper }>
						<label>Cities</label>
						<CityInput />
						{ this.renderList() }
					</div>
				</div>
			</div>
		);
	}
}

const styles = {
	headerWrapper: {
		flexDirection: 'row',
	},
};

const mapStateToProps = state => {
	return {
		cities: state.cities
	};
};

export default connect(
	mapStateToProps,
	null
)(CityList);