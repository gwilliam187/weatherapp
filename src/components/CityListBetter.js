import React from 'react';
import { connect } from 'react-redux';
import { FixedSizeList as List } from 'react-window';
import AutoSizer from 'react-virtualized-auto-sizer';

import CityInput from './CityInput';
import CityButton from './CityButton';

class CityListBetter extends React.Component {
	renderRow = ({ index, style }) => {
		const city = this.props.cities[index];

		return <div style={style}><CityButton city={ city } key={ city._id } /></div>
	}

	renderList() {
		if(this.props.cities.length > 0) {
			return (
				<AutoSizer>
					{({ height, width }) => (
						<List
								height={ height }
								width={ width }
								itemCount={ this.props.cities.length }
								itemSize={ 70 }>
							{ this.renderRow }	
						</List>
					)}
				</AutoSizer>
			);
		} else {
			return (
				<div className='default'>City list empty</div>
			);
		}
	}

	render() {
		return (
			<div className='col-12 card city-list'>
				<div className='card-body'>
					<div style={ styles.headerWrapper }>
						<label>Cities</label>
						<CityInput />
					</div>
					<div style={ styles.listWrapper }>
						{ this.renderList() }
					</div>
				</div>
			</div>
		);
	}
}

const styles = {
	headerWrapper: {
		flexDirection: 'row',
	},
	listWrapper: {
		height: '100%',
	},
};

const mapStateToProps = state => {
	return {
		cities: state.cities
	};
};

export default connect(
	mapStateToProps,
	null
)(CityListBetter);