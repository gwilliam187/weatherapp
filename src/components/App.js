import React from 'react';
import { ToastContainer, toast, Flip } from 'react-toastify';
import { connect } from 'react-redux';
import 'react-toastify/dist/ReactToastify.css';

import { loadCities, updateCityToUser, addUser } from '../actions';
import { fetchWeathersForSelectedCities } from '../actions/weatherActions';
import UserList from './UserList';
import UserInput from './UserInput';
import UserBar from './UserBar';
import CityList from './CityList';
import CityInput from './CityInput';
import WeatherDisplay from './WeatherDisplay';
import DiffDisplay from './DiffDisplay';
import './custom-style.css';

class App extends React.Component {
	componentDidMount(){
		//this.props.addUser("steven_klarens");
		this.props.loadCities();
		// this.props.updateCityToUser();
	}

	componentDidUpdate() {
		this.props.fetchWeathersForSelectedCities();
	}

	renderSidebar() {
		if(this.props.selectedUser) {
			return (
				<div className='col-lg-4'>
					<div className='row'>
						<UserBar />
						<CityList />
						<CityInput />
					</div>
				</div>
			);
		} else {
			return (
				<div className='col-12 d-flex'>
					<div className='row flex-grow-1 align-self-end'>
						<UserInput />
					</div>
				</div>
			);
		}
	}

	renderWeatherDisplay() {
		if(this.state.cityList.length === 1) {
			return (
				<div className='row'>
					<div className='col-12 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		} else if(this.state.cityList.length === 2) {
			return (
				<div className='row'>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		}
	}

	renderDiffDisplay() {
		if(this.state.tempList.length === 2) {
			return (
				<div className='row'>
					<div className='col-12'>
						<DiffDisplay tempList={this.state.tempList} windList={this.state.windList} />
					</div>
				</div>
			);
		}
	}

	render() {
		return (
			<div className="container app">
				<div className='row flex-grow-1'>
					{ this.renderSidebar() }
					<div className='col-lg-8'>
						<WeatherDisplay />
						<DiffDisplay />
					</div>
				</div>
				<ToastContainer
					toastClassName='custom-toast'
					position="top-right"
					autoClose={3000}
					hideProgressBar
					newestOnTop={false}
					closeOnClick
					rtl={false}
					pauseOnVisibilityChange
					draggable
					pauseOnHover
					transition={Flip}
				/>
			</div>
		);
	}
}

const mapStateToProps = state => {
	return {
		selectedCities: state.selectedCities,
		selectedUser: state.selectedUser
	};
};

const mapDispatchToProps = {
	fetchWeathersForSelectedCities, 
	loadCities, 
	updateCityToUser,
	addUser
};

export default connect(mapStateToProps, mapDispatchToProps)(App);