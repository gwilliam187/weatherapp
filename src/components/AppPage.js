import React from 'react';
import { toast } from 'react-toastify';
import { connect } from 'react-redux';
import 'react-toastify/dist/ReactToastify.css';
import { Redirect } from 'react-router-dom';

import { createDB, setDB } from '../actions/rxdbActions';
import { loadCities, loadTrees } from '../actions';
import { fetchWeathersForSelectedCities } from '../actions/weatherActions';
import { selectObject } from '../actions/selectedObjectActions';
import RegionBar from './RegionBar';
import UserInput from './UserInput';
import UserBar from './UserBar';
import CityList from './CityList';
import CityInput from './CityInput';
import WeatherDisplay from './WeatherDisplay';
import DiffDisplay from './DiffDisplay';
import TreeList from "./TreeList";
import './custom-style.css';

toast.configure({
	autoClose: 3000
});

class AppPage extends React.Component {
	async componentDidMount() {
		if(this.props.selectedRegion) {
			// const db = await createDB(this.props.selectedUser, this.props.selectedRegion)
			// this.props.setDB(db)
			// this.props.loadCities(db);
			// this.props.loadTrees(db);
		}
	}

	componentDidUpdate() {
		// this.props.fetchWeathersForSelectedCities();
	}

	renderSidebar() {
		return (
			<div className='col-lg-4'>
				<div className='row'>
					<CityList />
					<TreeList />
					<CityInput />
				</div>
			</div>
		);
	}

	renderWeatherDisplay() {
		if(this.state.cityList.length === 1) {
			return (
				<div className='row'>
					<div className='col-12 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		} else if(this.state.cityList.length === 2) {
			return (
				<div className='row'>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
					<div className='col-md-12 col-lg-6 pb-3'>
						<WeatherDisplay />
					</div>
				</div>
			);
		}
	}

	renderDiffDisplay() {
		if(this.state.tempList.length === 2) {
			return (
				<div className='row'>
					<div className='col-12'>
						<DiffDisplay tempList={this.state.tempList} windList={this.state.windList} />
					</div>
				</div>
			);
		}
	}

	renderCityDisplay() {
		return (
			<CityList />
		);
	}

	renderTreeDisplay() {
		return (
			<TreeList />
		);
	}

	renderDisplay() {
		if(this.props.selectedObject === 'cities') { return this.renderCityDisplay() }
		else if(this.props.selectedObject === 'trees') return this.renderTreeDisplay()
		else return <div></div>
	}

	getCityNavClass() {
		let css = 'btn btn-link nav-link ';
		if(this.props.selectedObject === 'cities') {
			css += 'active';
		}
		return css;
	}

	getTreeNavClass() {
		let css = 'btn btn-link nav-link ';
		if(this.props.selectedObject === 'trees') {
			css += 'active';
		}
		return css;
	}

	render() {
		if(this.props.selectedRegion) {
			return (
				<div className="container app align-items-start">
					<div className='row flex-grow-1'>
						<RegionBar />
						<div className='col-12 nav nav-pills mb-2'>
							<li className='nav-item'>
								<button 
										onClick={() => { this.props.selectObject('cities') }} 
										className={ this.getCityNavClass() }>Cities</button>
							</li>
							<li className='nav-item'>
								<button 
										onClick={() => { this.props.selectObject('trees') }}
										className={ this.getTreeNavClass() }>Trees</button>
							</li>
						</div>
						{ this.renderDisplay() }
					</div>
				</div>
			);
		} else {
			return <Redirect to='/' />;
		}
	}
}

const mapStateToProps = state => {
	return {
		selectedRegion: state.selectedRegion,
		selectedObject: state.selectedObject,
	};
};

const mapDispatchToProps = {
	fetchWeathersForSelectedCities, 
	loadCities, 
	loadTrees,
	createDB,
	setDB,
	selectObject,
};

export default connect(mapStateToProps, mapDispatchToProps)(AppPage);