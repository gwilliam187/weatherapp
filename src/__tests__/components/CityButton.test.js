import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { CityButton } from '../../components/CityButton';

describe('<CityButton />', () => {
  let wrapper;
  const mockSelectCity = jest.fn();
  const mockUpdateCitiesErrorMessage = jest.fn();
  const mockOnClick = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
    	<CityButton 
    		selectCity={mockSelectCity} 
    		updateCitiesErrorMessage={mockUpdateCitiesErrorMessage} />
    );
  });

  test('Handles onClick', () => {
  	wrapper.find('button').simulate('click');
  	expect(mockSelectCity).toHaveBeenCalled();
  	expect(mockUpdateCitiesErrorMessage).toHaveBeenCalled();
  });

	test('Renders the component', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});