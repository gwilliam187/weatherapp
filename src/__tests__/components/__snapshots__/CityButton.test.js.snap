// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityButton /> Renders the component 1`] = `
<div
  className="py-2"
>
  <button
    className="btn btn-outline-secondary btn-lg city-button"
    onClick={[Function]}
  />
</div>
`;
